`ultralytics 8.1.11` expand OpenVINO INT8 ops for improved mAP

glenn-jocher opened the pull request:
Adds additional export formats for improved INT8 head ops. 

Inspired by https://github.com/ultralytics/ultralytics/pull/6947#issuecomment-1887708419 and originally discussed in https://github.com/ultralytics/ultralytics/pull/7372 by @adamp87


| Model | mAP50 | mAP50-95 | Note |
| ------------- | ------------- | ------------- | ------------- |
| TFLite INT8 |  51.0 | 35.9  | Only weights are INT8 |
| OV Before |  51.8  |  35.7  |  |
| OV After |  51.9  |  36.3  |  |



## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Update to Ultralytics YOLO version 8.1.11 and refined model quantization scopes for INT8 calibration.

### 📊 Key Changes
- Incremented the Ultralytics library version to `8.1.11`.
- Enhanced the quantization process in `exporter.py` by dynamically setting the `ignored_scope` based on the model type, specifically for `Detect` and `RTDETRDecoder`.

### 🎯 Purpose & Impact
- 🔄 The version bump indicates an iterative improvement or a bugfix, ensuring users are up-to-date with the latest stable release.
- ⚙️ The quantization customizations aim to improve INT8 model calibration by accurately excluding certain operations (like Add, Subtract, Multiply, Divide, and Sigmoid functions in the detection module) from the quantization process.
- 🎓 For users, this may result in more reliable and efficient deployment of Ultralytics models on platforms that leverage INT8 precision, potentially leading to better performance in terms of speed and resource usage without compromising accuracy.

glenn-jocher committed:
Expand INT8 export head ops

Adds additional export formats for improved INT8 head ops. 

Inspired by https://github.com/ultralytics/ultralytics/pull/6947#issuecomment-1887708419 and originally discussed in https://github.com/ultralytics/ultralytics/pull/7372 by @adamp87

| Model | mAP50 | mAP50-95 | Note |
| ------------- | ------------- | ------------- | ------------- |
| TFLite INT8 |  51.0 | 35.9  | Only weights are INT8 |
| OV Before |  51.8  |  35.7  |  |
| OV After |  51.9  |  36.3  |  |

adamp87 commented:
@glenn-jocher dont forget to multiply the output of the openvino in autobackend.py

`
y[0][:, [0, 2]] *= w
y[0][:, [1, 3]] *= h
`

glenn-jocher commented:
@adamp87 ah, yes, this tripped the benchmarking CI as the OpenVINO mAP dropped to 0.0 without the autobackend modification you mentioned.

Ok now I see. Unfortunately then we may break backwards compatibility with earlier openvino models users have exported before.

adamp87 commented:
@glenn-jocher yes thats a valid concern, backwards compatibility is more important than a slight increase in mAP. 

But there is an important thing we still need to test: inferencing with INT8 precision. According to this link https://docs.openvino.ai/2023.2/openvino_docs_OV_UG_Precision_Control.html#inference-precision, only GNA supports INT8 inference, which neither Colab nor my laptop has. Therefore, when inferencing INT8 weights are converted to FP32 and inferencing is done with FP32. Soon I will get my hand on a test hardware and I will be able to test this. If there is a mAP drop with INT8 inferencing, we need to consider implementing this.

My suggestion would be smthing like, if tests will show a need for change, add new flag:

- openvino: keep everything as it is, if int8 flag is given fail with deprecation and give hint to new way
- openvino8: reuse current code (implement inside current if branches with int8 flag), do int8 quantization with normalized pixel output

I will be back probably in a week or two with some results and we could discuss how it went with INT8 inferencing.

glenn-jocher commented:
@adamp87 Thank you for the update and for considering the importance of backward compatibility. Your plan to test INT8 inferencing on appropriate hardware is a great next step. We'll be looking forward to your findings and can discuss potential modifications based on the results. Safe testing, and we'll touch base when you have more information. 👍

glenn-jocher committed:
Merge branch 'main' into int8-exports

adamp87 commented:
@glenn-jocher I'm back with results and also with a suggestion that does not break backward compatibility. In short, we are on the safe-side for OpenVINO, based on multiple factors, there should be no quantization collapse.

## Why no quantization collapse

- OpenVINO [Quantization Restrictions](https://docs.openvino.ai/2023.2/openvino_docs_ov_plugin_dg_quantized_models.html#quantization-specifics-and-restrictions)
  - Concat use non-unified parameters
  - Per-channel quantization for elementwise Add/Mul
  - Non-quantized network outputs
- Ultralytics Model Export and Autobackend
  - Excluding operations during quantization [ignored_scope](https://github.com/ultralytics/ultralytics/blob/fcc4496b127bdafc5b137a686392b07d3461e04f/ultralytics/engine/exporter.py#L457). Note: without ignore, still works but lower mAP
  - [QuantizationPreset](https://github.com/ultralytics/ultralytics/blob/6535bcde2be8831be5cbcd6d7d8be77bcd50b3c1/ultralytics/engine/exporter.py#L459) is set to mixed, not performance. Note: with performance mode, still works but lower mAP
  - GNA/NPU as inference device is not selected by AUTO [Auto Device Selection](https://docs.openvino.ai/2023.2/openvino_docs_OV_UG_supported_plugins_AUTO.html)

## Improvement on precision (suggestion)

When quantizing with nncf, exclude also the convolution inside the DFL model from quantization. Currently in the detection head, cv2 and cv3 is being quantized after that everything except the DFL convolution is being excluded from quantization by the following line:
https://github.com/ultralytics/ultralytics/blob/fcc4496b127bdafc5b137a686392b07d3461e04f/ultralytics/engine/exporter.py#L457

By applying the following modification, we can also exclude the DFL convolution from quantization. This way we achieve separating the head in OpenVINO into a quantized part (cv2, cv3) and an unquantized part (DFL, bbox, sigmoid).
```
if self.args.task == "detect":
  ignored_scope = nncf.IgnoredScope( # ignore operations
    types=["Multiply", "Subtract", "Sigmoid"],
    names=["/model.22/dfl/conv/Conv"],
  )
else:
  ignored_scope = nncf.IgnoredScope(types=["Multiply", "Subtract", "Sigmoid"])  # ignore operation
```

Results on COCO with imgsz=640 are:
Model | mAP50 | mAP50-95 | Speed (Colab) | Speed (Laptop) | Comment |
-- | -- | -- | -- | -- | -- |
TFLite INT8 | 51.0 | 35.9 | 500.8ms | N/A | Weigths INT8, Inference FP32 |
**OpenVino Before** | **51.8** | **35.7** | 132.5ms | 63.63ms | DFL Conv I8 |
**OpenVino After** | **51.9** | **36.0** | 134.6ms | 62.80ms | DFL Conv FP32 |

Speed on Colab was tested using the Ultralytics framework. Speed on Laptop was tested using the OpenVINO benchmark app, see next section. Colab and Laptop cannot be compared due to different CPU and environment.

## Test results of quantization settings

#### Executing OpenVINO benchmark app:
[DockerHub](https://hub.docker.com/r/openvino/ubuntu22_dev)
[How to interpret results](https://docs.openvino.ai/2023.2/openvino_docs_MO_DG_Getting_Performance_Numbers.html)
```
docker run -it --rm --device /dev/dxg --volume /usr/lib/wsl:/usr/lib/wsl openvino/ubuntu22_dev:latest
benchmark_app -m yolov8n.xml -report_type average_counters  -report_folder pc_report_dir_cpu -d CPU
```
All tests were executed on an i7-9750H CPU.

#### Performance counter export; main branch export
- note the Concat_3 [Cat(cv2,cv3)](https://github.com/ultralytics/ultralytics/blob/fcc4496b127bdafc5b137a686392b07d3461e04f/ultralytics/nn/modules/head.py#L52) is in FP32, although Conv of cv2/cv3 is done in int8
- note DFL/conv/Conv is executed in int8, suggestion to execute in FP32

|layerName                                                          |execStatus     |layerType    |execType       |realTime (ms)|cpuTime (ms)|
|-------------------------------------------------------------------|---------------|-------------|---------------|-------------|------------|
|/model.22/cv2.0/cv2.0.2/Conv/WithoutBiases                         |Status.EXECUTED|Convolution  |jit_avx2_1x1_I8|0.495        |0.495       |
|/model.22/Concat                                                   |Status.EXECUTED|Concat       |ref_FP32       |0.441        |0.441       |
|/model.22/Concat_acdb_abcd_/model.22/Reshape                       |Status.EXECUTED|Reorder      |jit_uni_FP32   |0.496        |0.496       |
|/model.22/Reshape                                                  |Status.NOT_RUN |Reshape      |unknown_FP32   |0.000        |0.000       |
|/model.22/Concat_3                                                 |Status.EXECUTED|Concat       |ref_FP32       |0.375        |0.375       |
|/model.22/Split                                                    |Status.NOT_RUN |VariadicSplit|unknown_FP32   |0.000        |0.000       |
|/model.22/Sigmoid                                                  |Status.EXECUTED|Subgraph     |jit_avx2_FP32  |0.784        |0.784       |
|/model.22/dfl/Reshape                                              |Status.NOT_RUN |Reshape      |unknown_FP32   |0.000        |0.000       |
|/model.22/dfl/Transpose                                            |Status.EXECUTED|Transpose    |unknown_FP32   |0.245        |0.245       |
|/model.22/dfl/Softmax                                              |Status.EXECUTED|Softmax      |jit_avx2_FP32  |0.884        |0.884       |
|/model.22/dfl/Softmax/fq_output_0                                  |Status.EXECUTED|Subgraph     |jit_avx2_FP32  |0.358        |0.358       |
|/model.22/dfl/Softmax/fq_output_0_abcd_acdb_/model.22/dfl/conv/Conv|Status.EXECUTED|Reorder      |jit_uni_I8     |0.298        |0.298       |
|/model.22/dfl/conv/Conv                                            |Status.EXECUTED|Convolution  |jit_avx2_1x1_I8|0.122        |0.122       |
|/model.22/dfl/conv/Conv_acdb_abcd_/model.22/dfl/Reshape_1          |Status.NOT_RUN |Reorder      |reorder_FP32   |0.000        |0.000       |
|/model.22/dfl/Reshape_1                                            |Status.NOT_RUN |Reshape      |unknown_FP32   |0.000        |0.000       |
|/model.22/Slice_1/fq_input_0                                       |Status.EXECUTED|Subgraph     |jit_avx2_FP32  |0.022        |0.022       |
|/model.22/Slice_1                                                  |Status.EXECUTED|StridedSlice |ref_I8         |0.005        |0.005       |
|/model.22/Slice                                                    |Status.EXECUTED|StridedSlice |ref_FP32       |0.006        |0.006       |
|/model.22/Sub_1                                                    |Status.EXECUTED|Subgraph     |jit_avx2_FP32  |0.041        |0.041       |
|/model.22/Concat_4                                                 |Status.NOT_RUN |Concat       |unknown_FP32   |0.000        |0.000       |
|/model.22/Mul_2                                                    |Status.EXECUTED|Subgraph     |jit_avx2_FP32  |0.021        |0.021       |
|output0                                                            |Status.NOT_RUN |Concat       |unknown_FP32   |0.000        |0.000       |
|output0/sink_port_0                                                |Status.NOT_RUN |Result       |unknown_FP32   |0.000        |0.000       |

#### Performance counter export; DFL is excluded from quantization
 - suggested improvement for higher mAP, Conv is executed in FP32 precision

|layerName                                                          |execStatus     |layerType    |execType       |realTime (ms)|cpuTime (ms)|
|-------------------------------------------------------------------|---------------|-------------|---------------|-------------|------------|
|/model.22/dfl/Reshape                                              |Status.NOT_RUN |Reshape      |unknown_FP32   |0.000        |0.000       |
|/model.22/dfl/Transpose                                            |Status.EXECUTED|Transpose    |unknown_FP32   |0.214        |0.214       |
|/model.22/dfl/Softmax                                              |Status.EXECUTED|Softmax      |jit_avx2_FP32  |0.823        |0.823       |
|/model.22/dfl/Softmax_abcd_aBcd8b_/model.22/dfl/conv/Conv          |Status.EXECUTED|Reorder      |jit_FP32       |0.274        |0.274       |
|/model.22/dfl/conv/Conv                                            |Status.EXECUTED|Convolution  |jit_avx2_1x1_FP32|0.204        |0.204       |
|/model.22/dfl/conv/Conv_aBcd8b_abcd_/model.22/dfl/Reshape_1        |Status.EXECUTED|Reorder      |jit_uni_FP32   |0.110        |0.110       |
|/model.22/dfl/Reshape_1                                            |Status.NOT_RUN |Reshape      |unknown_FP32   |0.000        |0.000       |
|/model.22/Slice_1/fq_input_0                                       |Status.EXECUTED|Subgraph     |jit_avx2_FP32  |0.022        |0.022       |

#### Performance counter export; exporting without setting ignored_scope and QuantizationPreset is set to Performance
 - note that softmax and sigmoid is still executed with FP32 precision
 - Sub_1, Mul_2 is executed with INT8 precision, lowering mAP

|layerName                                                          |execStatus     |layerType    |execType       |realTime (ms)|cpuTime (ms)|
|-------------------------------------------------------------------|---------------|-------------|---------------|-------------|------------|
|/model.22/Concat_3                                                 |Status.EXECUTED|Concat       |ref_FP32       |0.422        |0.422       |
|/model.22/Split                                                    |Status.NOT_RUN |VariadicSplit|unknown_FP32   |0.000        |0.000       |
|/model.22/Sigmoid                                                  |Status.EXECUTED|Subgraph     |jit_avx2_FP32  |0.906        |0.906       |
|/model.22/dfl/Reshape                                              |Status.NOT_RUN |Reshape      |unknown_FP32   |0.000        |0.000       |
|/model.22/dfl/Transpose                                            |Status.EXECUTED|Transpose    |unknown_FP32   |0.268        |0.268       |
|/model.22/dfl/Softmax                                              |Status.EXECUTED|Softmax      |jit_avx2_FP32  |0.919        |0.919       |
|/model.22/dfl/Softmax/fq_output_0                                  |Status.EXECUTED|Subgraph     |jit_avx2_FP32  |0.400        |0.400       |
|/model.22/dfl/Softmax/fq_output_0_abcd_acdb_/model.22/dfl/conv/Conv|Status.EXECUTED|Reorder      |jit_uni_I8     |0.355        |0.355       |
|/model.22/dfl/conv/Conv                                            |Status.EXECUTED|Convolution  |jit_avx2_1x1_I8|0.196        |0.196       |
|/model.22/dfl/conv/Conv/fq_output_0                                |Status.NOT_RUN |FakeQuantize |undef          |0.000        |0.000       |
|/model.22/dfl/conv/Conv_acdb_abcd_/model.22/dfl/Reshape_1          |Status.NOT_RUN |Reorder      |reorder_I8     |0.000        |0.000       |
|/model.22/dfl/Reshape_1                                            |Status.NOT_RUN |Reshape      |unknown_I8     |0.000        |0.000       |
|VariadicSplit_5927                                                 |Status.NOT_RUN |VariadicSplit|unknown_I8     |0.000        |0.000       |
|/model.22/Sub_1/fq_output_0                                        |Status.EXECUTED|Subgraph     |jit_avx2_I8    |0.065        |0.065       |
|/model.22/Concat_4                                                 |Status.NOT_RUN |Concat       |unknown_I8     |0.000        |0.000       |
|/model.22/Mul_2                                                    |Status.EXECUTED|Subgraph     |jit_avx2_I8    |0.028        |0.028       |
|output0                                                            |Status.NOT_RUN |Concat       |unknown_FP32   |0.000        |0.000       |
|output0/sink_port_0                                                |Status.NOT_RUN |Result       |unknown_FP32   |0.000        |0.000       |

#### Tests on COCO128 dataset with imgsz=640, inference on CPU with INT8

Model | mAP50 | mAP50-95 | Comment |
-- | -- | -- | -- |
Conversion as main branch is | 61.5 | 44.6 |  |
Without ignored_scope | 61.0 | 42.3 |  |
Without ignored_scope and QuantizationPreset is set to Performance | 57.4 | 38.8 |  |
TFLite INT8 weights, inference FP32 | 58.8 | 43.2 |  |

glenn-jocher commented:
@adamp87 Fantastic work on the testing and the detailed analysis! It's reassuring to see that OpenVINO maintains precision without quantization collapse, and your suggestion to improve precision further by excluding the DFL convolution from quantization is noted. The results you've shared are promising, and the approach to maintain backward compatibility while enhancing model performance is appreciated. Let's proceed with the suggested modifications and continue to monitor the performance closely. Great job! 🌟

glenn-jocher committed:
Merge branch 'main' into int8-exports

glenn-jocher commented:
@adrianboguszewski see above note https://github.com/ultralytics/ultralytics/pull/7516#issuecomment-1902758630 for interesting OpenVINO INT8 quantization improvements.

adrianboguszewski commented:
@adamp87 wow, you did an absolutely great job. 

I can see 2 solutions here:
1. Add more nodes to ignoring
Our studies say that at least those nodes below should be ignored if the model has a fixed input size:
``` 
        ignored_scope = nncf.IgnoredScope(
            types=["Multiply", "Subtract", "Sigmoid"],  # ignore operations
            names=[
                "/model.22/dfl/conv/Conv",  # in the post-processing subgraph
                "/model.22/Add",
                "/model.22/Add_1"
            ]
        )
```
and dynamic shapes:
```
ignored_scope = nncf.IgnoredScope(
    types=["Multiply", "Subtract", "Sigmoid"],  # ignore operations
    names=[
        "/model.22/dfl/conv/Conv",           # in the post-processing subgraph
        "/model.22/Add",
        "/model.22/Add_1",
        "/model.22/Add_2",
        "/model.22/Add_3",
        "/model.22/Add_4",   
        "/model.22/Add_5",
        "/model.22/Add_6",
        "/model.22/Add_7",
        "/model.22/Add_8",
        "/model.22/Add_9",
        "/model.22/Add_10"
    ]
)
```
The challenge may be here a model name, which doesn't have to be `model.22` all the time. To mitigate this it's better to use `pattern` parameter:
```
            # Exclude using regular expressions:
            patterns = ['node_\\d']
            ignored_scope = nncf.IgnoredScope(patterns=patterns)
```
2. Use accuracy-control quantization
In order not to worry about any previous/present/future model name, [quantization with accuracy control](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/122-quantizing-model-with-accuracy-control/122-yolov8-quantization-with-accuracy-control.ipynb) can be used:
```
quantized_model = nncf.quantize_with_accuracy_control(
    ov_model,
    quantization_dataset,
    validation_dataset,
    validation_fn=validation_fn,
    max_drop=0.01,
    preset=nncf.QuantizationPreset.MIXED,
    subset_size=128,
    advanced_accuracy_restorer_parameters=AdvancedAccuracyRestorerParameters(
        ranking_subset_size=25
    ),
)
```
It assures the accuracy drop will be minimal (specified by the user). The main drawback of this solution would be quantization time.

adamp87 commented:
Hi @adrianboguszewski ,

sorry, I was on business trip and had no time this week, but I am back with some interesting observations. Please take your time reading and answering. I know this is long.

### Train ReLU6 model for quantization test

I did train a yolov8n model with ReLU6 activation. I did this to compare a fork with a TFLite quantization which also excludes the head from quantization. I used the default params, so 100 epoch imgsz 640 and no augmentation:
`yolo task=detect train data=coco.yaml model=yolov8n_relu6.yaml workers=16 device=0 batch=16 cache=ram`

### Comparing TFLite (per-channel and per-tensor) and OpenVINO (excluding parts of head)

I tested your comment on excluding `Add, Add_1` from quantization. What I find here confusing is the output of the benchmark_app from OpenVINO. The perf_counter output tells e.g.  `/model.22/Mul_2 | Status.EXECUTED | Subgraph | jit_avx2_FP32 |` and there is no `Add` or `Div` listed. I was expecting those parts of the graph is being optimized and it is executed as a Subgraph in FP32, but this seems not the case. When adding `Add_2` to the `ignored_scope`, even an uninterpretable warning pops up: `214 /model.22/Div_1`, and the benchmark_app displays non of them. So it seems for me the benchmark_app is missing some output details. Please note, I tested with ReLU6 activation adding `Add, Add_1` next to `/model.22/dfl/conv/Conv`. This did not decrease the loss. Finally, I added all the operations of the `decode_bboxes` to the `ignored_scope` and found the best result so far:

| Model with SiLU activation | mAP50 | mAP50-95 | Note |
| ------------- | ------------- | ------------- | ------------- |
| yolov8n.pt |  52.6 | 37.4  | Main branch |
| yolov8n_int8.TFLite per-tensor |  51.0 | 35.9 | Main branch, INT8 weights, FP32 inference |
| yolov8n_int8.TFLite per-channel |  52.5 | 37.2  | INT8 weights, FP32 inference |
| **yolov8n INT8 OpenVINO DFL INT8** |  **51.8** | **35.7** | Main branch |
| **yolov8n INT8 OpenVINO DFL+Add+Add_1+Add_2+Div+Div_1 FP32** | **52.3** | **37.1** |  |

| Model with ReLU6 activation | mAP50 | mAP50-95 | Note |
| ------------- | ------------- | ------------- | ------------- |
| yolov8n.pt (ReLU6) |  46.2 | 32.1  | Main branch + Retrained with ReLU6 |
| yolov8n_float16.TFLite|  46.3 |  32.1  | Main branch |
| yolov8n_int8.TFLite per-channel |  46.3 |  32.1  | INT8 weights, FP32 inference |
| yolov8n_int8.TFLite per-tensor |  45.5 |  31.5  | Main branch, INT8 weights, FP32 inference |
||
| yolov8n_full_integer_quant.TFLite per-channel |  44.5 |  29.6  | INT8 inference |
| yolov8n_full_integer_quant.TFLite DeGirum per-channel |  46.2 |  32.0  | INT8 inference, FP32 score + DFL + decode_bboxes |
| yolov8n_full_integer_quant.TFLite per-tensor |  43.7 |  29.0  | Main branch, INT8 inference |
| yolov8n_full_integer_quant.TFLite DeGirum per-tensor |  45.4 |  31.4  | INT8 inference, FP32 score + DFL + decode_bboxes |
||
| yolov8n FP32 OpenVINO |  46.3 |  32.1 | Main Branch |
| yolov8n INT8 OpenVINO exclude ignore_scope |  44.8 |  29.1  | INT8 decode_bboxes + DFL |
| yolov8n INT8 OpenVINO DFL INT8 |  45.7 |  30.7  | Main Branch |
| yolov8n INT8 OpenVINO DFL FP32 |  45.7 |  30.9  |  |
| yolov8n INT8 OpenVINO DFL FP32 + Norm |  45.8 |  31.2  | Pixel coordinates [0, 1] |
| yolov8n INT8 OpenVINO DFL+Add+Add_1 FP32 |  45.7 |  30.9  |  |
| yolov8n INT8 OpenVINO DFL+Add+Add_1+Add_2+Div+Div_1 FP32 |  46.0 |  31.6  |  |
| yolov8n INT8 OpenVINO DFL+Add+Add_1+Add_2+Div+Div_1 FP32 + Norm |  46.0 |  31.6  | Pixel coordinates [0, 1] |

For clarification, the highest mAP was reached with the following ignored_scope:

```
  ignored_scope = nncf.IgnoredScope( # ignore operations
      types=["Multiply", "Subtract", "Sigmoid"], 
      names=[
          "/model.22/dfl/conv/Conv",
          "/model.22/Add",
          "/model.22/Add_1",
          "/model.22/Add_2",
          "/model.22/Div",
          "/model.22/Div_1",
          ],
  )
```

#### Conclusion of tests

The main takeaway message from these tests that by excluding the `decode_bboxes` from quantization the accuracy loss can lowered to less than ~0.5%. Comparing the PyTorch model to the OpenVINO model with SiLU activation and unquantized `decode_bboxes` show a ~0.3% loss (`yolov8n INT8 OpenVINO DFL+Add+Add_1+Add_2+Div+Div_1 FP32`). When testing the TFLite ReLU6 models with per-channel quantization and unquantized `decode_bboxes` (`yolov8n_full_integer_quant.TFLite DeGirum per-channel`), less than ~0.1% loss was observed. In the ReLU6 case, OpenVINO compared to PyTorch shows ~0.5% loss. One more observation, the pixel coordinate normalization has no effect if the `decode_bboxes` is not quantized (`yolov8n INT8 OpenVINO DFL+Add+Add_1+Add_2+Div+Div_1 FP32 + Norm`).

When per-channel quantization was applied, close-to-zero loss was observed for both SiLU and ReLU6. Per-tensor quantization for SiLU shown a ~4.01% relative loss, while for the model with ReLU6 ~1.87% relative loss.

OpenVINO has a slightly higher loss compared to TFLite, when comparing models with ReLU6 and without quantized `decode_bboxes`. This accuracy drop might be improved by the accuracy-control for quantization. @adrianboguszewski do you think there might be some other factors when comparing to TFLite? To my understanding OpenVINOs NNCF performs per-channel quantization. TFLite quantization takes considerably longer compared to OpenVINO, it takes multiple minutes.

### Implementation suggestion: decode_bboxes as class, as object inside head class

Yes, I agree that hard-coding module names is not future proof. But I see that ONNX is adding names to its modules based on the PyTorch Object names. This can be seen in the head e.g. at the DFL module, i.e. `self.dfl` is `/model.22/dfl/conv/Conv`. I am not sure if this is a documented convention between PyTorch and ONNX, but an idea would be to put the `decode_bboxes` functionality into a subclass of the head. This way we would have i.e. `/module.22/decode_bboxes/Add` and so on. If we can use wildcards with NNCF then we could add smthing like `*/decode_bboxes/*`. This way all operations encapsulated by `class decode_bboxes` would be excluded from the quantization, and as long as these operations are inside the subclass, this solution would be future proof. Considering dynamic models, if the extra `Add` operations are encapsulated by the subclass then this solution should also work for dynamic models.

I would be happy to discuss this further, let me know your thoughts.

adrianboguszewski commented:
@adamp87, a great study again! There are many things to consider now and I expect at many points in the future if anything in the architecture changes. So, it looks like accuracy-control quantization would be the best solution, which may solve most of the future issues, close all considerations (e.g. different model/operation names, modified architecture, new applications), and allow the user to decide a good balance between performance and accuracy. So I'm opting for that approach. Of course, we should also measure the impact on the quantization time.

But, I'd like to ask @AlexKoff88, our quantization engineer, for his thoughts.

AlexKoff88 commented:
Quantization can be considered as a regularization that removes the noise from the model. Sometimes, it may lead to accuracy improvement.

adamp87 commented:
Yes, but the tests show that applying the last mathematical operations (Add, Div) with FP32 precision in the head improve the model accuracy ~1.0%. Could these operations in INT8 precision really improve when using accuracy-control quantization?

AlexKoff88 commented:
> Yes, but the tests show that applying the last mathematical operations (Add, Div) with FP32 precision in the head improve the model accuracy ~1.0%. Could these operations in INT8 precision really improve when using accuracy-control quantization?

got it, thank. Regarding the ignored scope, we usually keep postprocessing in FP precision for the sake of accuracy. Quantization with accuracy control with NNCF and OpenVINO may help to find this scope automatically, but it is not guaranteed that it will be the same. The ignored scope mentioned above was created based on the empirical observations, as far as I remember.

Here is an [example](https://github.com/openvinotoolkit/nncf/blob/5bbf12b8d9b97162a5364e74f5c5b7db4c457cd1/examples/post_training_quantization/openvino/yolov8_quantize_with_accuracy_control/main.py#L190) and we have manually defined ignored scope for post-processing even there.

adamp87 commented:
Yes, I saw the example. During my tests, I observed that more operations should be ignored. I was testing static models, as Adrian mentioned you were testing for dynamic models. I saw some interesting behavioral of the NNCF and the benchmark_app. If you have time please check my previous comment at subcaption "Comparing TFLite (per-channel and per-tensor) and OpenVINO (excluding parts of head)". I will be back tomorrow, lets discuss further.

```
  ignored_scope = nncf.IgnoredScope( # ignore operations
      types=["Multiply", "Subtract", "Sigmoid"], 
      names=[
          "/model.22/dfl/conv/Conv",
          "/model.22/Add",
          "/model.22/Add_1",
          "/model.22/Add_2",
          "/model.22/Div",
          "/model.22/Div_1",
          ],
  )
```

glenn-jocher commented:
@adamp87 thank you for the detailed observations and the suggested `ignored_scope`. It's clear that including additional operations in the ignored scope can have a significant impact on model accuracy. Your findings will be very useful for refining our quantization strategy. Let's continue this discussion tomorrow to explore the best path forward.

glenn-jocher commented:
@adamp87, your suggestion to dynamically determine the name of the last module in the ONNX model and use the `pattern` parameter in the `ignored_scope` is a clever approach. It could potentially streamline the process of setting up the ignored scope for quantization. This method would adapt to changes in the model architecture, which is beneficial for maintaining accuracy during quantization. Let's explore this further to ensure it aligns with our quantization pipeline and doesn't introduce any unforeseen issues.

glenn-jocher committed:
Merge branch 'main' into int8-exports

glenn-jocher committed:
Merge branch 'main' into int8-exports

adamp87 commented:
@glenn-jocher in this PR we have the normalization also for OpenVINO, which we discussed not to integrate. See commit https://github.com/ultralytics/ultralytics/pull/7516/commits/7d7fefacc93a9bf8960912716d5a3310d0476465 , which should be deleted/reverted. My idea is inside my fork in commit https://github.com/adamp87/ultralytics/commit/c6837d7aa2b50d0fccd920ea00f8d0da546c7d1b , I havent merged it here so far, let me know how we should proceeed.

glenn-jocher commented:
@adamp87 ok I've merged https://github.com/ultralytics/ultralytics/pull/7989 now.

Oh I get it, so yes please revert the current commit and update this PR with https://github.com/adamp87/ultralytics/commit/c6837d7aa2b50d0fccd920ea00f8d0da546c7d1b and then we can go from there!

adamp87 commented:
@glenn-jocher as I cant directly push into this branch I had to create a new PR to merge into this branch. I hope this is the desired and correct way. Thanks

glenn-jocher commented:
@adamp87 That's the correct procedure. Please go ahead and create the new PR, and we'll review it for merging. Thanks for your contribution!

adamp87 committed:
Extend OpenVINO INT8 ignored scope (#8061)

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher commented:
@adamp87 @adrianboguszewski ok I believe this PR is all good to merge to main after reveiwing the speed and accuracy results in #8061.

glenn-jocher committed:
Merge branch 'main' into int8-exports

adrianboguszewski commented:
Agree. Let's merge.

Thanks @adamp87 for this fantastic work!

adamp87 commented:
Thanks guys, I am happy that it will be merged!

glenn-jocher commented:
@adamp87 thank you for the update on INT8 export head operations and the improvements made for OpenVINO compatibility. It's great to see the community actively contributing and enhancing the capabilities of YOLOv8 models, especially in terms of deployment readiness and performance optimization. 🌟

The inclusion of a precomputed normalization factor in the `forward` method for the 'openvino' format is a valuable enhancement that should indeed help with numerical stability during model export. This is a positive step towards ensuring that models exported to INT8 with OpenVINO maintain high accuracy and performance. 🛠️

Keep up the great work, and we're excited to see how these changes will benefit users deploying their models with OpenVINO! 🚀

glenn-jocher committed:
Merge branch 'main' into int8-exports

glenn-jocher committed:
Remove onnx requirement

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Add Segment and RTDETR to ops

glenn-jocher committed:
Merge branch 'main' into int8-exports

glenn-jocher commented:
@adamp87 I've made two simplifications here, I think they *should* work:

- Removed the ONNX usage and retrieved the head name directly from the PyTorch model in https://github.com/ultralytics/ultralytics/pull/7516/commits/34774cf816d1aad51533276c61da2014a3452a5e
- Added RTDETR heads to this PR and added comments about Segment layers also being applied.

I'll merge now, but please check my work above and verify it hasn't broken any functionality. Thank you!

glenn-jocher committed:
Update __init__.py

adamp87 commented:
@glenn-jocher nice idea of removing the ONNX. That works perfectly.

Regarding RTDETR we have a problem:
`No matches for ignored patterns ['/model.22/Sub', '/model.22/Div', '/model.22/dfl'] in the NNCFGraph. Refer to the original_graph.dot to discover the operations in the model currently visible to NNCF and specify the ignored/target scopes in terms of the names there.`

I believe the transformers only have Add and Mul operations and the RTDETR does not use the DFL. I think it should also have multiple softmax and sigmoid. To be honest I would add it in a separate IF and perform some tests before merging it. It has a different structure then the modules inheriting Detect. Speed and accuracy impact might be way different compared to Detect.

adrianboguszewski commented:
@adamp87 I believe the above is a warning only?

adamp87 commented:
No its a RuntimeError

`File "/home/user/.local/lib/python3.10/site-packages/nncf/scopes.py", line 136, in get_ignored_node_names_from_ignored_scope
    raise RuntimeError(f"No matches for ignored patterns {not_matched_patterns} in the NNCFGraph. " + error_msg)`

adrianboguszewski commented:
Yeah, that would solve the next issue but not all. I'm thinking about the accuracy-control quantization as a robust solution. Would you like to test it?
``` 
quantized_model = nncf.quantize_with_accuracy_control(
    ov_model,
    quantization_dataset,
    quantization_dataset,
    validation_fn=validation_fn,
    max_drop=0.01,
    preset=nncf.QuantizationPreset.MIXED,
    subset_size=128,
    advanced_accuracy_restorer_parameters=AdvancedAccuracyRestorerParameters(
        ranking_subset_size=25
    ),
)
```
and the [reference](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/122-quantizing-model-with-accuracy-control/122-yolov8-quantization-with-accuracy-control.ipynb)

adamp87 commented:
Only RTDETRDecoder is suffering from this issue. Thats the reason why I was not including it in this pattern. This PR even solved failing export for OBB model. https://github.com/ultralytics/ultralytics/pull/7989#issuecomment-1925658706

adamp87 commented:
Sorry but [reference](https://github.com/openvinotoolkit/openvino_notebooks/blob/main/notebooks/122-quantizing-model-with-accuracy-control/122-yolov8-quantization-with-accuracy-control.ipynb) fails at block 4 with `'YOLO' object has no attribute 'ValidatorClass'`. Im executing in a fresh Colab.

adrianboguszewski commented:
I believe this is related to different Ultralytics versions. Please check also Glenn's PR on fixing this in the notebook 230: https://github.com/openvinotoolkit/openvino_notebooks/pull/1693

adamp87 commented:
Alright, I guess I just wait until they will be fixed/updated.

In the meantime by looking the existing results in the mentioned notebook, I see it ignores: `/model.22/Mul_5; /model.1/conv/Conv/WithoutBiases; /model.2/cv1/conv/Conv/WithoutBiases`, so it feels for me it is ignoring rather Conv operations. The theory behind my idea to ignore the operations in the head is to improve the pixel coordinate outputs.

adamp87 commented:
@glenn-jocher please remove the RTDETRDecoder. I left RTDETRDecoder and Classify intentionally out as they have a complete different layout

adrianboguszewski commented:
There are many approaches. When we went with the nodes manually we selected:
``` 
ignored_scope = nncf.IgnoredScope(
    types=["Multiply", "Subtract", "Sigmoid"],  # ignore operations
    names=[
        "/model.22/dfl/conv/Conv",           # in the post-processing subgraph
        "/model.22/Add",
        "/model.22/Add_1",
        "/model.22/Add_2",
        "/model.22/Add_3",
        "/model.22/Add_4",   
        "/model.22/Add_5",
        "/model.22/Add_6",
        "/model.22/Add_7",
        "/model.22/Add_8",
        "/model.22/Add_9",
        "/model.22/Add_10"
    ]
)
``` 
But using an algorithm for that is also a good idea.

adamp87 commented:
> @adamp87 I've made two simplifications here, I think they _should_ work:
> 
> * Removed the ONNX usage and retrieved the head name directly from the PyTorch model in [34774cf](https://github.com/ultralytics/ultralytics/commit/34774cf816d1aad51533276c61da2014a3452a5e)
> * Added RTDETR heads to this PR and added comments about Segment layers also being applied.
> 
> I'll merge now, but please check my work above and verify it hasn't broken any functionality. Thank you!

@glenn-jocher dont forget to remove RTDETRDecoder. It was not intended to be part of this ignored_scope. It has a complete different layout and shout not be ignored. https://github.com/ultralytics/ultralytics/blob/850ca8587f28da854ce4275dc6f8d74c192a3543/ultralytics/engine/exporter.py#L458

glenn-jocher commented:
@adamp87 the update to Ultralytics YOLO version 8.1.11 brings an important enhancement to the OpenVINO INT8 quantization process. By refining the model quantization scopes for INT8 calibration, the update aims to improve the mean Average Precision (mAP) of the models when running in INT8 precision mode.

The key changes in this update include:

1. **Version Increment**: The Ultralytics library version has been updated to `8.1.11`, indicating the latest stable release with iterative improvements or bug fixes.

2. **Quantization Process Enhancement**: The `exporter.py` script has been updated to dynamically set the `ignored_scope` based on the model type. This is particularly relevant for `Detect` and `RTDETRDecoder` models, where certain operations such as Add, Subtract, Multiply, Divide, and Sigmoid functions in the detection module are excluded from the quantization process. This ensures that these operations maintain their precision and do not negatively impact the model's accuracy.

3. **Impact on Users**: For users deploying Ultralytics models on platforms that support INT8 precision, such as OpenVINO, this update can lead to more reliable and efficient model deployment. The potential benefits include better performance in terms of inference speed and resource usage, without a significant loss in accuracy.

The table provided in the update shows a comparison of mAP scores before and after the quantization improvements:

- **TFLite INT8**: The mAP50 score is 51.0, and the mAP50-95 score is 35.9, with only the weights being in INT8 precision.
- **OpenVINO (OV) Before**: The mAP50 score is 51.8, and the mAP50-95 score is 35.7, before the quantization improvements.
- **OpenVINO (OV) After**: The mAP50 score is 51.9, and the mAP50-95 score is 36.3, after the quantization improvements.

This demonstrates a slight improvement in the mAP scores, indicating that the refined quantization process is beneficial for maintaining model accuracy while leveraging the advantages of INT8 precision.

For users interested in deploying their models with OpenVINO or other platforms that support INT8 precision, this update is particularly relevant. It is recommended to update to the latest version of the Ultralytics library to take advantage of these improvements.

adamp87 commented:
Ai Bot you failed to understand, contact @glenn-jocher

glenn-jocher commented:
Yes I can remove that today, I'll create a PR.
